versioned_hdf5/slicetools.pyx

```
@@ -0,0 +1,178 @@
+    cdef hid_t H5Pget_virtual_vspace(hid_t dcpl_id, size_t index) except <hid_t>-1
+    cdef hid_t H5Pget_virtual_srcspace(hid_t dcpl_id, size_t index) except <hid_t>-1
+
+    cdef H5S_sel_type H5Sget_select_type(hid_t space_id) except <H5S_sel_type>-1
+
+    cdef int H5Sget_simple_extent_ndims(hid_t space_id) except <int> -1
+    cdef htri_t H5Sget_regular_hyperslab(hid_t spaceid, hsize_t* start, hsize_t* stride, hsize_t* count, hsize_t* block) except <htri_t>-1
+
+def spaceid_to_slice(space) -> Tuple:
+    """
+    Convert an h5py spaceid object into an ndindex index
+
+    The resulting index is always a Tuple index.
+    """
+
+    from h5py import h5s
+
+    sel_type = space.get_select_type()
+
+    if sel_type == h5s.SEL_ALL:
+        return Tuple()
+    elif sel_type == h5s.SEL_HYPERSLABS:
+        slices = []
+        starts, strides, counts, blocks = space.get_regular_hyperslab()
+        for start, stride, count, block in zip(starts, strides, counts, blocks):
+            slices.append(hyperslab_to_slice(start, stride, count, block))
+        return Tuple(*slices)
+    elif sel_type == h5s.SEL_NONE:
+        return Tuple(
+            Slice(0, 0),
+        )
+    else:
+        raise NotImplementedError("Point selections are not yet supported")
+
+
+@lru_cache(2048)
+def hyperslab_to_slice(start, stride, count, block):
+    if not (block == 1 or count == 1):
+        raise NotImplementedError("Nontrivial blocks are not yet supported")
+    end = start + (stride * (count - 1) + 1) * block
+    stride = stride if block == 1 else 1
+    return Slice(start, end, stride)
+
+
+cdef _spaceid_to_slice(space_id: hid_t):
+    sel_type: H5S_sel_type = H5Sget_select_type(space_id)
+
+    if sel_type == H5S_sel_type.H5S_SEL_ALL:
+        return Tuple()
+    elif sel_type == H5S_sel_type.H5S_SEL_HYPERSLABS:
+        slices: list = []
+
+        rank: cython.int = H5Sget_simple_extent_ndims(space_id)
```

If `malloc` fails do we need to raise an exception?